### PR TITLE
Allow allocating new TXD's

### PR DESCRIPTION
**Pull request by TheNormalnij**
Part of #1792

### API:
```lua
local txdID = engineRequestTXD( "TXDNAME" )
engineFreeTXD( txdID )
```

### CODE REVIEW

**Comment by qaisjp** [comment]
It is typical to use elements for things like this.

You mentioned earlier that you were hesitant to reuse our txd elements here. Why's that?

Is it because engineLoadTXD (which returns a txd element) has actual texture data associated with it, and this one doesn't / is more like an "empty container"?

I don't really understand the semantics.

**Comment by TheNormalnij** [comment]
Because i can't use CClientTXD ("txd" element) class and I'm don't sure that a new class "allocated-txd" is a good idea.

**Comment by qaisjp** [comment]
> Because i can't use CClientTXD ("txd" element) class

Why not?

**Comment by TheNormalnij** [comment]
Because this id doesn't present a file. So i need put `if (m_iPoolID)` for all method's in CClientTXD class.

**Comment by qaisjp** [comment]
Okay. maybe others have ideas.

what can the txd ID be used for anyway>

**Comment by AlexTMjugador** [comment]
I think this function is somewhat like `engineRequestModel`, in the sense that it is just meant to return a TXD ID (slot) that can then be used for a model ID. At least from the scripting side, I can't justify the overhead of promoting the TXD ID to a fully qualified element, as the bare integer is all that is needed to work with it, and it wouldn't have any methods apart from requesting and freeing the TXD slot and associating it to a model ID. On the MTA code side, I think it would be better to follow the example of [CClientModel](https://github.com/multitheftauto/mtasa-blue/blob/master/Client/mods/deathmatch/logic/CClientModel.cpp) to keep things tidy.

**Comment by qaisjp** [comment]
Okay I'm convinced.

I still don't understand what TXD IDs are for. Please can someone ELI5?

**Comment by TheNormalnij** [comment]
@qaisjp i will explain.
GTA use `streamID` for dynamic loaded objects(TXD,DFF,COL,IFP,IPL...) to find them in IMG archives.
```c++
struct CStreamingInfo
{
    uint16 prevId;
    uint16 nextId;
    uint16 nextInImg;
    uint8  flg;
    uint8  archiveId;
    uint32 offsetInBlocks;
    uint32 sizeInBlocks;
    uint32 loadState;
}
CStreamingInfo (&CStreamingSA::ms_aInfoForModel)[26316] = *(CStreamingInfo(*)[26316])0x8E4CC0;
```
Info for objects stored into another arrays
```c++
// Array size is 20000
CBaseModelInfoSAInterface** CModelInfoSAInterface::ms_modelInfoPtrs = (CBaseModelInfoSAInterface**)0xA9B0C8;
```
This is array with pointers to models. Almost every model use same DFF as `modelID` (CJ, broken car parts use different loading).
TXD, COL, IFP stored into `CPool<Type>`
Models use field `txdID` as pointer(position in pool) to TXD and `animID` as pointer to IFP. GTA allocate them while read IDE. 
GTA put CStreamingInfo while read .IMG header by file type and find right slotID in pool by file name.

To find resource in .IMG from id use something like this:
```c++
streamID = POOL_TYPE_OFFSET + id
```
For TXD this:
```c++
streamID = 20000 + txdID
```

So we will use `txdID` for:
* Set another texture dictionary for model.
* Set file in .IMG for new allocated TXD.
* Or change streaming info for existing TXD's (HD mods)

**Comment by qaisjp** [comment]
Thank you very much! This explanation is very insightful.

File: Client/game_sa/CPoolsSA.cpp [hunk 1]
```diff
@@ -1393,14 +1393,17 @@ int CPointerNodeSingleLinkPoolSA::GetNumberOfUsedSpaces()
...
     CTextureDictonarySAInterface* pTxd = (*m_ppTxdPoolInterface)->Allocate();
     if (!pTxd)
         return -1;
 
+    typedef uint(__cdecl * Function_TxdGatHashFromName)(char* strName);
+    Function_TxdGatHashFromName getHash = (Function_TxdGatHashFromName)(0x53CF30);
+
     pTxd->usUsagesCount = 0;
-    pTxd->hash = 0;
+    pTxd->hash = getHash(&strTxdName.at(0));
```
Comment (Pirulax) [inline Client/game_sa/CPoolsSA.cpp]: I think this should be `.c_str()` instead?

File: Client/game_sa/CPoolsSA.h [hunk 2]
```diff
@@ -241,7 +241,7 @@ class CPoolsSA : public CPools
     void ResetPedPoolCount() { m_pedPool.ulCount = 0; }
     void InvalidateLocalPlayerClientEntity();
 
-    uint AddTextureDictonarySlot();
+    uint AddTextureDictonarySlot(std::string strTxdName);
```
Comment (Pirulax) [inline Client/game_sa/CPoolsSA.h]: perhaps use `const std::string&`, because you won't be `std::move`ing the value, so a const ref is fine

File: Client/sdk/game/CPools.h [hunk 3]
```diff
@@ -135,6 +135,6 @@ class CPools
     virtual void ResetPedPoolCount() = 0;
     virtual void InvalidateLocalPlayerClientEntity() = 0;
 
-    virtual uint AddTextureDictonarySlot() = 0;
+    virtual uint AddTextureDictonarySlot(std::string strTxdName) = 0;
```
Comment (Pirulax) [inline Client/sdk/game/CPools.h]: Same.

**Review by Pirulax — CHANGES_REQUESTED** [review]

File: Client/game_sa/CPoolsSA.h (hunk 2 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.h]: Thanks

File: Client/game_sa/CPoolsSA.cpp (hunk 1 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.cpp]: This is only draft and some API still in progress.

**Comment by TheNormalnij** [comment]
I think i will use [CClientModel class](https://github.com/multitheftauto/mtasa-blue/blob/master/Client/mods/deathmatch/logic/CClientModel.cpp) as manager for new TXD's. Because we don't need managers for every type.

**Comment by Pirulax** [comment]
If a TXD is considered a model in GTA as well, they sure, go ahead.

**Comment by TheNormalnij** [comment]
Ready

File: Client/game_sa/CPoolsSA.cpp [hunk 4]
```diff
@@ -1419,3 +1420,42 @@ int CPointerNodeSingleLinkPoolSA::GetNumberOfUsedSpaces()
 
     return iOut;
 }
+
+unsigned int CPoolsSA::AllocateTextureDictonarySlot(uint uiSlotId, std::string& strTxdName)
```
Comment (Pirulax) [inline Client/game_sa/CPoolsSA.cpp:1059]: Perhaps use `const char*` instead of `std string`?

File: Client/game_sa/CPoolsSA.cpp [hunk 5]
```diff
@@ -1419,3 +1420,42 @@ int CPointerNodeSingleLinkPoolSA::GetNumberOfUsedSpaces()
...
+
+void CPoolsSA::RemoveTextureDictonarySlot(uint uiTxdId)
+{
+    if (!(*m_ppTxdPoolInterface)->IsContains(uiTxdId))
+        return;
+
+    typedef uint(__cdecl * Function_TxdReleaseSlot)(uint uiTxdId);
+    Function_TxdReleaseSlot freeTxdSlot = (Function_TxdReleaseSlot)(0x731E90);
+
+    freeTxdSlot(uiTxdId);
```
Comment (Pirulax) [inline Client/game_sa/CPoolsSA.cpp]: `((Function_TxdReleaseSlot)(0x731E90))(uxTxdId)` is fine imho.

File: Client/game_sa/CPoolsSA.h [hunk 6]
```diff
@@ -72,6 +80,82 @@ class CPoolSAInterface
...
+            if (this->m_byteMap[index].bEmpty)
+            {
+                this->m_nFirstFree = index;
+                return index;
+            }
+            index++;
+        }
+
+        return -1;
+    };
```
Comment (Pirulax) [inline Client/game_sa/CPoolsSA.h]: Is there any reason to use `this->`?

File: Client/mods/deathmatch/logic/CClientModel.cpp [hunk 7]
```diff
@@ -125,4 +136,33 @@ void CClientModel::RestoreEntitiesUsingThisModel()
...
+}
+
+bool CClientModel::AllocateTXD(std::string &strTxdName)
+{
+    uint uiSlotID = g_pGame->GetPools()->AllocateTextureDictonarySlot(m_iModelID - MAX_MODEL_DFF_ID, strTxdName);
+    if (uiSlotID == -1)
+        return false;
+
+    return true;
+}
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/CClientModel.cpp]: return uiSlotID != -1;

File: Client/mods/deathmatch/logic/CClientModel.cpp [hunk 8]
```diff
@@ -125,4 +136,33 @@ void CClientModel::RestoreEntitiesUsingThisModel()
...
+
+bool CClientModel::DeallocateTXD(CModelInfo* pModelInfo)
+{
+    uint uiTextureDictonarySlotID = pModelInfo->GetModel() - MAX_MODEL_DFF_ID;
+
+    for (uint uiModelID = 0; uiModelID < MAX_MODEL_DFF_ID; uiModelID++)
+    {
+        CModelInfo* pModelInfo = g_pGame->GetModelInfo(uiModelID, true);
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/CClientModel.cpp:188]: What if `pModelInfo` is `nullptr`?

File: Client/mods/deathmatch/logic/CClientModel.h [hunk 9]
```diff
@@ -32,13 +33,17 @@ class CClientModel
...
-    void             RestoreEntitiesUsingThisModel();
-    void             SetParentResource(CResource* pResource) { m_pParentResource = pResource; }
-    CResource*       GetParentResource(void) const { return m_pParentResource; }
+    int                             GetModelID(void) const { return m_iModelID; };
+    eClientModelType                GetModelType(void) const { return m_eModelType; };
+    bool                            Allocate(ushort usParentID);
+    bool                            AllocateTXD(std::string &strTxdName);
+    bool                            Deallocate(void);
+    void                            SetParentResource(CResource* pResource) { m_pParentResource = pResource; }
+    CResource*                      GetParentResource(void) const { return m_pParentResource; }
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/CClientModel.h]: Why the extra space?

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 10]
```diff
@@ -2033,3 +2033,29 @@ bool CLuaEngineDefs::EngineRestreamWorld(lua_State* const luaVM)
     g_pClientGame->RestreamWorld();
     return true;
 }
+
+uint CLuaEngineDefs::EngineRequestTXD(lua_State* const luaVM, std::string strTxdName)
+{
+    if (strTxdName.size() > 24)
+        throw std::invalid_argument("Expected string size up to 24");
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp]: I'd suggest to change this to something like:
> TXD name length shoudn't be more than 24 characters.

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp [hunk 11]
```diff
@@ -2033,3 +2033,29 @@ bool CLuaEngineDefs::EngineRestreamWorld(lua_State* const luaVM)
...
+    return iModelID - 20000;
+}
+
+bool CLuaEngineDefs::EngineFreeTXD(uint txdID)
+{
+    std::shared_ptr<CClientModel> pModel = m_pManager->GetModelManager()->FindModelByID(20000 + txdID);
+    if (pModel && pModel->Deallocate())
+        return true;
+
+    return false;
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp]: `return pModel && pModel->Deallocate();`

**Review by Pirulax — DISMISSED** [review]

File: Client/mods/deathmatch/logic/CClientModel.cpp (hunk 8 above)
Comment (TheNormalnij) [reply Client/mods/deathmatch/logic/CClientModel.cpp:188]: `pModelInfo` can't be `nullptr` here because second arg for `GetModelInfo` is `true`

File: Client/game_sa/CPoolsSA.cpp (hunk 4 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.cpp:1059]: Any reason?

**Comment by patrikjuvonen** [comment]
Merge conflicts must be resolved.

**Comment by TheNormalnij** [comment]
Fixed errors after merge

File: Client/game_sa/CPoolsSA.h [hunk 12]
```diff
@@ -54,8 +62,76 @@ class CPoolSAInterface
...
+    {
+        m_byteMap[index].bEmpty = true;
+        m_byteMap[index].nId = 0;
+        if (index == m_nFirstFree)
+            --m_nFirstFree;
+    }
+
+    void Delete(uint index)
+    {
+        delete m_pObjects[index];
```
Comment (Pirulax) [inline Client/game_sa/CPoolsSA.h]: Are you sure about that?
Perhaps you meant to call the destructor?
There's no need to call `delete`.

File: Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.h [hunk 13]
```diff
@@ -78,6 +78,9 @@ class CLuaEngineDefs : public CLuaDefs
     static bool                                            EngineSetModelVisibleTime(std::string strModelId, char cHourOn, char cHourOff);
     static std::variant<bool, CLuaMultiReturn<char, char>> EngineGetModelVisibleTime(std::string strModelId);
 
+    static uint CLuaEngineDefs::EngineRequestTXD(lua_State* const luaVM, std::string strTxdName);
+    static bool CLuaEngineDefs::EngineFreeTXD(uint txdID);
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.h]:
```suggestion
    static uint EngineRequestTXD(lua_State* const luaVM, std::string strTxdName);
    static bool EngineFreeTXD(uint txdID);
```

File: Client/game_sa/CPoolsSA.h [hunk 14]
```diff
@@ -54,8 +62,76 @@ class CPoolSAInterface
...
+        return -1;
+    };
+
+    B* Allocate()
+    {
+        uint index = GetFreeSlot();
+        if (index == -1)
+            return nullptr;
+
+        return AllocateAt(index);
```
Comment (Pirulax) [inline Client/game_sa/CPoolsSA.h]: Improved version of the original code.
```suggestion
        m_nFirstFree++;          // Continue after the last allocated slot
        const auto sz = m_nSize; // Storing size to avoid reloads from memory - should help out the optimizer
        for (auto i = 0u; i < sz; i++) {
            const auto slot = (m_nFirstFree + i) % sz;
            const auto e    = &m_byteMap[slot];
            if (!e->bEmpty) {
                continue;
            }
            m_nFirstFree = slot;
            e->bEmpty = false;
            e->nId++;
            return &m_pObjects[slot];
        }
        return nullptr;
```

**Review by Pirulax — COMMENTED** [review]
Address these suggestions and I'll happily merge Sir.

File: Client/game_sa/CPoolsSA.h (hunk 14 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.h:98]: I suggest 
```c++
m_nFirstFree = slot;
```
after
```c++
 `e->nId++;` 
```

File: Client/game_sa/CPoolsSA.cpp (hunk 4 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.cpp:1059]: std::string& can be replaced with string_view here

File: Client/game_sa/CPoolsSA.h (hunk 14 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.h]: I suggest
```c++
m_nFirstFree = slot;
```
This helps skip original allocated items

Comment (Pirulax) [reply Client/game_sa/CPoolsSA.h]: Ah shit I forgot to add that. I've fixed the suggestion.
Thanks!

File: Client/game_sa/CPoolsSA.cpp (hunk 4 above)
Comment (Pirulax) [reply Client/game_sa/CPoolsSA.cpp:1059]: Are you sure?
IIRC a few lines down `strTxdName.resize(24)` is called

File: Client/game_sa/CPoolsSA.h (hunk 12 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.h]: Hmm... I wrote this code with a little experience 3 years ago.
Now i have no clue why this line doesn't break building.
This doesn't look right at all.

File: Client/game_sa/CPoolsSA.h [hunk 15]
```diff
@@ -17,6 +17,14 @@
 
 #define INVALID_POOL_ARRAY_ID 0xFFFFFFFF
 
+struct CTextureDictonarySAInterface
+{
+    RwTexDictionary* rwTexDictonary;
+    unsigned short   usUsagesCount;
+    unsigned short   usParentIndex;
+    unsigned int     hash;
+};
```
Comment (TheNormalnij) [inline Client/game_sa/CPoolsSA.h]: Should not be declared here i think

File: Client/game_sa/CPoolsSA.h (hunk 12 above)
Comment (TheNormalnij) [inline Client/game_sa/CPoolsSA.h]:
```suggestion
```

File: Client/game_sa/CPoolsSA.cpp (hunk 4 above)
Comment (TheNormalnij) [reply Client/game_sa/CPoolsSA.cpp:1059]: You are right

**Comment by Xenius97** [comment]
I've tested `engineSetModelTXDID` and `engineRequestTXD`, look good to me, but I noticed a possible memory leak. When you restart the resource, the requested ID won't be freed anymore (maybe if I manually call `engineFreeModel`). The same applies to `engineRequestTXD` and `engineFreeTXD`.

https://youtu.be/VoTWDzTKSjs
(Some default skins not working, See #3032)

**Comment by Pirulax** [comment]
Is there a good reason as to why `engineRequestTXD` should exist instead of moving it into `engineRequestModel`?
@TheNormalnij
I personally don't see anything code wise that would prevent us from merging the 2 functions together.

> but I noticed a possible memory leak. 

It's working as intended I believe.

I've looked at the video, and it seems like the texture only gets loaded after `setElementModel` is called once again after linking the img, right?

**Comment by ghost** [comment]
Well, kinda off-topic but setElementModel is big memory leak itself, see #2672

**Comment by TheNormalnij** [comment]
> Is there a good reason as to why `engineRequestTXD` should exist instead of moving it into `engineRequestModel`?

Function name "engineRequestModel" confuse if it can allocate a textures dictionary.
engineRequestTXD has a little bit different allocation logic, so i preferred extract logic in the function.

I would rather have two small functions that a big one by interface segregation principle.

**Comment by Pirulax** [comment]
Okay that's fair enough.

**Review by Pirulax — APPROVED** [review]
lgtm

**Comment by Pirulax** [comment]
Don't forget to add the Wiki pages
